Add mpio support for iscsi

tobad357 opened the pull request:
This allows the iscsi volume to check if a iscsi device belongs to a mpio device
If it does belong to the device then we make sure we mount the mpio device instead of
the raw device. 
The code is based on the current FibreChannel volume support for mpio

example
/dev/disk/by-path/iqn-example.com.2999 -> /dev/sde
Then we check
/sys/block/[dm-X]/slaves/xx
until we find the [dm-X] containing /dev/sde and mount it

Additional work that can be done in future
1. Add multiple portal support to iscsi
2. Move the FibreChannel volume provider to use the code that has been extracted

markturansky commented on pkg/volume/iscsi/iscsi_util.go:
```diff
@@ -98,6 +98,7 @@ func (util *ISCSIUtil) MakeGlobalPDName(iscsi iscsiDisk) string {
 }
 
 func (util *ISCSIUtil) AttachDisk(b iscsiDiskBuilder) error {
+	deviceUtil := b.deviceUtil
```
why the need for a local variable?  You can reference `b.deviceUtil` throughout this method and not require the local var.

tobad357 replied:
Leftover from some earlier code.. will clear up with any other comments

saad-ali commented on pkg/volume/util/device_util_linux.go:
```diff
@@ -0,0 +1,86 @@
+	//return just the last part
+	if len(parts) == 3 && strings.HasPrefix(parts[1], "dev") {
+		return parts[2], nil
+	}
```
Instead of repeating the same block of code before and after the `EvalSymlinks` you can just do it once after. `EvaSymLinks` will just return the original path if it is not a symlink. `EvaSymLinks("/dev/sda")`, for example, will return `"/dev/sda"`

saad-ali commented on pkg/volume/util/device_util_linux.go:
```diff
@@ -0,0 +1,86 @@
+		for _, f := range dirs {
+			name := f.Name()
+			if strings.HasPrefix(name, "dm-") {
+				if _, err1 := io.Lstat(sysPath + name + "/slaves/" + disk); err1 == nil {
```
Why `os.Lstat(...)`? Why not just `os.IsExist(...)`?

rootfs replied:
this is for unit test

saad-ali replied:
> this is for unit test
 
I don't follow. `IsExist(...)` could be stubbed out just as easily as `Lstat(...)`.

rootfs replied:
[os.IsExist](https://golang.org/pkg/os/#IsExist) doesn't help us check path existence.

saad-ali commented on pkg/volume/util/device_util_linux.go:
```diff
@@ -0,0 +1,86 @@
+// +build linux
+
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
```
Update copyright year on new files.

saad-ali commented on pkg/volume/util/device_util_linux.go:
```diff
@@ -0,0 +1,86 @@
+// +build linux
```
Is there anything here that wouldn't work for `darwin` (i.e. Mac)? Also create an "unsupported" version to gracefully handle unsupported builds (see `pkg/volume/util/fs_unsupported.go`, for example).

markturansky replied:
No, not supported on OSX. the path /sys/block, for example, does not exist.  We need the _unsupported file.

saad-ali commented on pkg/volume/util/device_util_linux.go:
```diff
@@ -0,0 +1,86 @@
+	"strings"
+)
+
+//FindMultipathDeviceForDevice given a device name like /dev/sdx, find the devicemapper parent
```
nit: Please add a space between `//` and the comment.

saad-ali commented on pkg/volume/util/device_util_linux_test.go:
```diff
@@ -0,0 +1,138 @@
+	return nil
+}
+
+func TestFindMultipathDeviceForDevice(t *testing.T) {
```
Thanks for writing unit tests!

saad-ali commented on pkg/volume/util/io_util.go:
```diff
@@ -0,0 +1,52 @@
+
+//IoUtil is a util for common IO operations
+//it also backports certain operations from golang 1.5
+type IoUtil interface {
```
`pkg/volume/fc/fc_util.go` defines `ioHandler` which does essentially the same thing. Could you consolidate that with this.

tobad357 replied:
My problem is I don't have any access to FC which means I can't test any changes
Can I do a follow-up pull request which can be tested separately?

rootfs replied:
if that's the way to go, please make sure opening an issue to track it.

saad-ali replied:
> Can I do a follow-up pull request which can be tested separately?

Yes. Just create a bug to follow up and put a TODO comment with the issue # here saying that this code is duplicate and it should be consolidated with `ioHandler` defined in `pkg/volume/fc/fc_util.go`.

saad-ali commented on pkg/volume/util/device_util.go:
```diff
@@ -0,0 +1,34 @@
+//it also backports certain operations from golang 1.5
+type DeviceUtil interface {
+	FindMultipathDeviceForDevice(disk string) string
+	FindDevicesForMultipathDevice(disk string) []string
```
I don't see where in the code `FindDevicesForMultipathDevice(...)` is actually used.

tobad357 replied:
It was needed before some changes landed to iscsi.
Shall I leave it in as a utility function or remove?

rootfs replied:
please address this comment

saad-ali replied:
Please remove unused code.

saad-ali commented on pkg/volume/iscsi/iscsi.go:
```diff
@@ -189,7 +193,8 @@ func (b *iscsiDiskBuilder) SetUpAt(dir string, fsGroup *int64) error {
 	*iscsiDisk
-	mounter mount.Interface
+	mounter    mount.Interface
+	deviceUtil ioutil.DeviceUtil
```
Looks like `DeviceUtil` only ends up getting consumed through `iscsiDiskBuilder`, so let's not add it to the `iscsiDiskCleaner` if it's not needed.

tobad357 commented:
Thank you @saad-ali I will get to these over the weekend.

rootfs commented on pkg/volume/util/device_util_linux.go:
```diff
@@ -0,0 +1,86 @@
+	return make([]string, 0, 0)
+}
+
+//FindDeviceForVirtual Find the underlaying disk for a linked path such as /dev/disk/by-path/XXXX or /dev/mapper/XXXX
```
comment doesn't match function name

rootfs commented:
@tobad357 I don't have an environment to test mpio but if you can help get a screenshot of a successful deployed pod and iscsi mount on the minion, that'll help. Thanks.

tobad357 commented:
@rootfs please see image of the kube registry mounted to a mpio mpath. I additionally have tested to delete the rc and make sure it unmounts and logs out properly
<img width="1407" alt="mpio_iscsi_kubernetes_screenshot" src="https://cloud.githubusercontent.com/assets/608607/13852887/49731dd0-ec9f-11e5-80a1-ce17d68e66cf.png">

rootfs commented:
@tobad357 can you rebase (instead of pull) against master branch?

tobad357 commented:
@rootfs Sorry bit late here so managed to muck it up. Working on it now

tobad357 commented:
@rootfs Seems like my tree is in quite a mess and/or my git-fu isn't strong enough. Do you mind if I close this pull request and open a new one with a new branch?

tobad357 commented:
@rootfs and @saad-ali I believe I managed to undo my mess. Please see last two commits.
They should fix all comments except

1. I haven't changed the Fibrechannel code as don't have a Fibrechannel env to test in. I would be happy to make the changes in a separate pull request that can be reviewed by itself

2. I've still left the FindDevicesForMultipathDevice in the code, it's not used and can be removed if you prefer

If your happy with the changes then I will squash

rootfs commented:
Some nits.

Please also update `example/iscsi/README.md` and include your config example, package requirement, and screenshot.

Rest LGTM.

saad-ali commented:
Sorry I haven't been able to take a look at this. Thanks for making the changes. Reviewing now.

saad-ali commented on pkg/volume/iscsi/iscsi.go:
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2015 The Kubernetes Authors All rights reserved.
+Copyright 2016 The Kubernetes Authors All rights reserved.
```
Sorry, I meant please change the copyright year from 2015 in the new files to 2016. Leave the existing files as is.

saad-ali commented on pkg/volume/util/device_util_linux.go:
```diff
@@ -0,0 +1,82 @@
+
+// FindDeviceForPath Find the underlaying disk for a linked path such as /dev/disk/by-path/XXXX or /dev/mapper/XXXX
+// will return sdX or hdX etc, if /dev/sdX is passed in then sdX will be returned
+func (handler *deviceHandler) FindDeviceForPath(path string) (string, error) {
```
Does this need to a public method? It is only used internally right?

tobad357 committed:
This is an update that allows the iscsi volume to check if a iscsi device belongs to a mpio device
If it does belong to the device then we make sure we mount the mpio device instead of
the raw device.

Heuristics
Login into /dev/disk/by-path/iqn-example.com.2999 -> /dev/sde
Check if sde existsin in /sys/block/[dm-X]/slaves/xx
If it does mount /dev/[dm-x] which will look like /dev/mapper/mpiodevicename in mount

examples/iscsi has more details

tobad357 commented:
@saad-ali 
I believe I have cleaned up the code with regards to the comments you gave.
I've also updated the example docs to mention the multipath support

The only question I have is that io_util disappeared recently (can't find in which commit).
I've re-added it as I use it for mocking all the IO functions.
I'm not sure if there is a better way of doing it?

saad-ali commented on pkg/volume/util/io_util.go:
```diff
@@ -0,0 +1,48 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
```
Since `io_util.go` is new, please use current copyright year.

saad-ali commented:
LGTM, a couple of minor comments remaining.

> The only question I have is that io_util disappeared recently (can't find in which commit). I've re-added it as I use it for mocking all the IO functions. I'm not sure if there is a better way of doing it?

I'm not sure what you mean. `io_util.go` is being added by you, I don't believe it existed before.

A reminder to address this comment:
> Yes. Just create a bug to follow up and put a TODO comment with the issue # here saying that this code is duplicate and it should be consolidated with `ioHandler` defined in `pkg/volume/fc/fc_util.go`.

Once you're done please squash the commits and change the commit message to something useful.

tobad357 commented:
@saad-ali Minor nits fixed and commits squashed and message updated

saad-ali commented:
Thanks @tobad357! LGTM

saad-ali commented:
@k8s-oncall This will need to be manually merged on green due to touching a blocked doc file. CC @bgrant0607

bgrant0607 commented:
@saad-ali I don't see any docs touched in this PR. examples are not included.

saad-ali commented:
> @saad-ali I don't see any docs touched in this PR. examples are not included.

I was under the impression that it was the modification to `examples/iscsi/README.md` that was triggering the `Adding label:do-not-merge because PR changes docs prohibited to auto merge` message from the merge-bot.

I'll remove the label and see if the bot readds it.

saad-ali commented:
@k8s-bot unit test this, issue: #24528

saad-ali commented:
@k8s-bot unit test this, issue: #24579
